common/rocksdb_env_s3.cpp

```diff
@@ -0,0 +1,432 @@
+/// Copyright 2018 Pinterest Inc.
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+/// http://www.apache.org/licenses/LICENSE-2.0
+
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+
+//
+// @author ghan (ghan@pinterest.com)
+//
+
+#include "common/rocksdb_env_s3.h"
+
+#include <string>
+
+#include "common/rocksdb_glogger/rocksdb_glogger.h"
+#include "gflags/gflags.h"
+#include "glog/logging.h"
+#include "rocksdb/status.h"
+
+DECLARE_bool(s3_direct_io);
+
+namespace rocksdb {
+
+// open a file for sequential reading
+Status S3Env::NewSequentialFile(const std::string& fname,
+                                std::unique_ptr<SequentialFile>* result,
+                                const EnvOptions& options) {
+  assert(s3_util_ != nullptr);
+  result->reset();
+
+  // We read first from local storage and then from S3.
```

Comment:
Could the file have changed in s3 but not locally?

Reply:
all the files used in the bakcup/restore are immutable.